File: apps/web/lib/assistant_personalize.ts
// 小爱定制化快捷提问（随用户读经特征变化）

import { getLastRead } from './reading';
import {
  chipUserQuestion,
  SCENES,
  type AssistantScene,
} from './assistant_scenes';

export interface AssistantChip {
  label: string;
  mode: string;
  scene: AssistantScene;
  q: string;
}

function chip(
  label: string,
  scene: AssistantScene,
  q: string,
): AssistantChip {
  return { label, scene, mode: SCENES[scene].mode, q };
}

export function personalizedAssistantChips(opts: {
  ref?: string;
  dailyVerseRef?: string;
  dailyTheme?: string;
  streak?: number;
}): AssistantChip[] {
  const anchor = opts.ref || opts.dailyVerseRef || '';
  const refLabel = anchor || undefined;
  const chips: AssistantChip[] = [];

  if (anchor) {
    chips.push(
      chip('经文背景', 'chat_explain', chipUserQuestion('解释经文', refLabel)),
      chip('生活应用', 'chat_apply', chipUserQuestion('生活应用', refLabel)),
      chip('预备查经', 'chat_study', chipUserQuestion('预备查经', refLabel)),
      chip('译本对照', 'chat_compare', chipUserQuestion('译本对照', refLabel)),
    );
  } else {
    chips.push(
      chip(
        '今日默想',
        'chat_apply',
        `根据今日经文${opts.dailyTheme ? `（主题：${opts.dailyTheme}）` : ''}，请给我 3 个适合个人的默想问题。`,
      ),
      chip('生活应用', 'chat_apply', chipUserQuestion('生活应用', refLabel)),
      chip('信仰问答', 'chat_understand', '作为读经初学者，请用浅显的中文解释「因信称义」是什么意思。'),
      chip('预备查经', 'chat_study', chipUserQuestion('预备查经', refLabel)),
    );
    // 有续读位置时仍给平行 pill，但不使用「续读导读」
    if (getLastRead()) {
      chips.push(chip('解释经文', 'chat_explain', chipUserQuestion('解释经文', refLabel)));
    }
  }

  if ((opts.streak ?? 0) >= 7) {
    chips.push(
      chip(
        '坚持鼓励',
        'chat_apply',
        `我已连续读经 ${opts.streak} 天，请根据这段属灵旅程给我一段鼓励与下一步建议。`,
      ),
    );
  }

  const seen = new Set<string>();
  return chips.filter((c) => {
    if (seen.has(c.label)) return false;
    seen.add(c.label);
    return true;
  }).slice(0, 6);
}

File: apps/mobile/lib/features/assistant/assistant_personalize.dart
/// 小爱定制化快捷提问（随用户读经特征变化）。
library;

import 'assistant_scenes.dart';
import 'models.dart';

class AssistantChip {
  const AssistantChip({
    required this.label,
    required this.scene,
    required this.q,
  });
  final String label;
  final AssistantScene scene;
  final String q;

  String get mode => scene.mode;

  AssistantMode get assistantMode =>
      AssistantMode.fromId(scene.mode) ?? AssistantMode.explain;
}

AssistantChip _chip(String label, AssistantScene scene, String q) =>
    AssistantChip(label: label, scene: scene, q: q);

/// 空态 / 输入区快捷 chip：有锚定经文、连续读经天数、今日主题时更贴身。
List<AssistantChip> personalizedAssistantChips({
  String? ref,
  String? dailyVerseRef,
  String? dailyTheme,
  int streak = 0,
  bool hasLastRead = false,
}) {
  final anchor = (ref != null && ref.isNotEmpty)
      ? ref
      : (dailyVerseRef != null && dailyVerseRef.isNotEmpty ? dailyVerseRef : '');
  final refLabel = anchor.isNotEmpty ? anchor : null;
  final chips = <AssistantChip>[];

  if (anchor.isNotEmpty) {
    chips.addAll([
      _chip(
        '经文背景',
        AssistantScene.chatExplain,
        chipUserQuestion('解释经文', ref: refLabel),
      ),
      _chip(
        '生活应用',
        AssistantScene.chatApply,
        chipUserQuestion('生活应用', ref: refLabel),
      ),
      _chip(
        '预备查经',
        AssistantScene.chatStudy,
        chipUserQuestion('预备查经', ref: refLabel),
      ),
      _chip(
        '译本对照',
        AssistantScene.chatCompare,
        chipUserQuestion('译本对照', ref: refLabel),
      ),
      _chip(
        '并列观点',
        AssistantScene.chatViewpoints,
        chipUserQuestion('并列观点', ref: refLabel),
      ),
    ]);
  } else {
    final themeBit =
        (dailyTheme != null && dailyTheme.isNotEmpty) ? '（主题：$dailyTheme）' : '';
    chips.addAll([
      _chip(
        '今日默想',
        AssistantScene.chatApply,
        '根据今日经文$themeBit，请给我 3 个适合个人的默想问题。',
      ),
      _chip(
        '生活应用',
        AssistantScene.chatApply,
        chipUserQuestion('生活应用', ref: refLabel),
      ),
      _chip(
        '信仰问答',
        AssistantScene.chatUnderstand,
        '作为读经初学者，请用浅显的中文解释「因信称义」是什么意思。',
      ),
      _chip(
        '预备查经',
        AssistantScene.chatStudy,
        chipUserQuestion('预备查经', ref: refLabel),
      ),
    ]);
    if (hasLastRead) {
      chips.add(
        _chip(
          '解释经文',
          AssistantScene.chatExplain,
          chipUserQuestion('解释经文', ref: refLabel),
        ),
      );
    }
  }

  if (streak >= 7) {
    chips.add(
      _chip(
        '坚持鼓励',
        AssistantScene.chatApply,
        '我已连续读经 $streak 天，请根据这段属灵旅程给我一段鼓励与下一步建议。',
      ),
    );
  }

  final seen = <String>{};
  return chips
      .where((c) {
        if (seen.contains(c.label)) return false;
        seen.add(c.label);
        return true;
      })
      .take(6)
      .toList();
}
